### PR TITLE
Feature/firebase upload test

### PR DESCRIPTION
**Pull request by rugeli**
Problem
=======
What is the problem this work solves, including
Added tests to firebase recipe upload 

Solution
========
What I/we did to solve this problem
created 3 test files for each class in `DBRecipeHandler`: `CompositionDoc`, `ObjectDoc`, and `DBRecipeHandler` 

with @meganrm 

## Type of change
Please delete options that are not relevant.

* new tests for #127 

Steps to Verify:
----------------
1. `conda activate autopack`
2. `pytest cellpack`

Keyfiles (delete if not relevant):
-----------------------
1. test_comp_doc.py
2. test_object_doc.py
3. test_db_recipe_handler.py
4. mock_db.py

### CODE REVIEW

File: cellpack/tests/test_comp_doc.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    composition_doc.check_and_replace_references(
+        objects_to_path_map, references_to_update, mock_db
+    )
+    assert composition_doc.as_dict()["object"] == "firebase:objects/test_id"
+    assert composition_doc.as_dict()["regions"] == {
+        "interior": [{"object": "firebase:objects/test_id", "count": 1}]
+    }
+
+
+def test_composition_oc_should_write_with_no_existing_doc():
```
Comment (mogres) [inline cellpack/tests/test_comp_doc.py:185]:
```suggestion
def test_composition_doc_should_write_with_no_existing_doc():
```

File: cellpack/tests/test_db_recipe_handler.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
+from cellpack.autopack.DBRecipeHandler import DBRecipeHandler
+from cellpack.tests.mocks.mock_db import MockDB
+
+mock_db = MockDB({})
+
+
+def test_is_nested_list():
+    assert DBRecipeHandler.is_nested_list([]) is False
+    assert DBRecipeHandler.is_nested_list([[], []]) is True
+    assert DBRecipeHandler.is_nested_list([[1, 2], [3, 4]]) is True
```
Comment (mogres) [inline cellpack/tests/test_db_recipe_handler.py:10]: Possible edge case:
```
[ 1, [1,2,3]]
```
Should this count as a nested list?

File: cellpack/tests/test_db_recipe_handler.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+
+def test_prep_data_for_db():
+    input_data = {
+        "bounding_box": [[0, 0, 0], [1000, 1000, 1]],
+        "positions": [[(0.0, 0.0, 0.0)], [(0.0, 50.0, 0.0), (43.26, -25.07, 0.0)]],
+        "max_jitter": [1, 1, 0],
+    }
+    converted_data = {
+        "bounding_box": {"0": [0, 0, 0], "1": [1000, 1000, 1]},
```
Comment (mogres) [inline cellpack/tests/test_db_recipe_handler.py:20]: Does firebase not handle numeric keys in dictionaries? I will need a walkthrough of Firebase at some point 😅

**Review by mogres — APPROVED** [review]
This looks great, thanks for adding the tests! A lot of work put into making cellPACK recipes available on the cloud!

File: cellpack/autopack/DBRecipeHandler.py [hunk 4]
```diff
@@ -202,8 +226,8 @@ def should_write(self, db, recipe_data):
                         ignore_type_in_groups=[tuple, list],
                     )
                     if not difference:
```
[the remark this answers is not in the document]
Comment (rugeli) [reply cellpack/autopack/DBRecipeHandler.py:228]: ###

File: cellpack/tests/mocks/mock_db.py [hunk 5]
```diff
@@ -1,23 +1,71 @@
+from unittest.mock import Mock
+
+
 class MockDB(object):
     def __init__(self, data) -> None:
         for index, name in enumerate(data):
             obj = data[name]
```
[the remark this answers is not in the document]
Comment (rugeli) [reply cellpack/tests/mocks/mock_db.py:7]: Yeah we can definitely initialize `obj` as `self.obj`. This obj is used to mock firestore classes, which we only used as a parameter placeholder and return True in tests, but we did used it in `FirebaseHandler` for object type checking

File: cellpack/tests/test_db_recipe_handler.py (hunk 2 above)
Comment (rugeli) [reply cellpack/tests/test_db_recipe_handler.py:10]: Thanks! you are right, it failed with this edge case. This function is mainly checking value's for `bounding_box`, `representation` and `position`. We could make it more robust for future use

File: cellpack/tests/test_db_recipe_handler.py (hunk 3 above)
Comment (rugeli) [reply cellpack/tests/test_db_recipe_handler.py:20]: yeah firebase has some limitations on data types that can be uploaded, we need to convert the data to be "firebase-uploadable" before uploading and revert it back to `cellpack-packable` after downloading. (painful process...)

**Review by meganrm — APPROVED** [review]

**Comment by mogres** [comment]
Could we use the `Squash and Merge` option while merging PRs? This will reduce the number of commits that get added to the main branch.

**Comment by rugeli** [comment]
> Could we use the `Squash and Merge` option while merging PRs? This will reduce the number of commits that get added to the main branch.

Sure, good idea. l'll find a way to redo the merge and squash the commits this time. Thank you!

**Comment by mogres** [comment]
> > Could we use the `Squash and Merge` option while merging PRs? This will reduce the number of commits that get added to the main branch.
> 
> Sure, good idea. l'll find a way to redo the merge and squash the commits this time. Thank you!

I don't think we need to (or can even) do it for this branch. Just something to keep in mind for future PRs :)
